packages/embark-test-runner/src/reporter.js

```diff
@@ -55,7 +55,14 @@ class EmbarkSpec extends Base {
       if (self.txDetails) {
         self.embarkEvents.request('contracts:contract', receipt.className, (contract) => {
           if (contract) {
-            self.contracts.push(contract);
+            let index = self.contracts.findIndex(c => c.className === contract.className);
+            // It's possible to deploy the same contract multiple times per test, so we need
+            // to make sure we replace the existing one with the new one.
+            if (index > -1) {
+              self.contracts = self.contracts.map(c => c.className === contract.className ? contract : c);
+            } else {
+              self.contracts.push(contract);
+            }
```

Comment:
Theoretically, we can always run the `.map()` and take the penalty of always iterating over all contracts, even if not necessary. I don't this is too much of a performance bottle-neck (unless we're dealing with hundreds of thousands of contracts in memory).

Reply:
With a `for...of` loop here we can avoid traversing the array twice and also stop the iteration with a `break` statement.

Reply:
I believe `findIndex()` is breaking out early as well.

Reply:
The following is what I had in mind:

```
if (contract) {
  let push = true;
  let index = 0;
  for (const c of contracts) {
    if (c.className === contract.className) {
      self.contracts[index] = contract;
      push = false;
      break;
    }
    index++;
  }
  if (push) self.contracts.push(contract);
  self.addressToContract = getAddressToContract(self.contracts, self.addressToContract);
}
```

If it's legit for the same `className` to appear 2+ times in `self.contracts` then we need to make the substitution more than once and can simply drop the `break` statement.

Reply:
I understand. In my comment I tried to express that we could probably just always `.map()` (for convenience, as it's just a single line), as we're most likely not dealing with a lot of contracts that could cause a performance problem at this point.

So while your proposal is absolutely correct, I think it'd make things look even more complex. Keep in mind that we're talking micro optimizations here.

Reply:
True, not a huge win.

I just had another thought. Since you already have the array index and the logic to find it is the same logic used in the map function, could we do:

```suggestion
              self.contracts[index] = contract;
```